src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,457 @@
+      }
+      if (line.startsWith("-")) {
+        return LineType.CHUNK_DEL;
+      }
+      if (line.startsWith(" ") || line.isEmpty()) {
+        return LineType.CHUNK_EQL;
+      }
+    } else {
+      if (line.startsWith("--- ")) {
+        return LineType.OLD_FILE;
+      }
+      if (line.startsWith("+++ ")) {
+        return LineType.NEW_FILE;
+      }
+      if (line.startsWith("diff --git ")) {
+        return LineType.GIT_HEADER;
+      }
+      if (line.startsWith("rename from ")) {
+        return LineType.RENAME_FROM;
+      }
+      if (line.startsWith("rename to ")) {
+        return LineType.RENAME_TO;
+      }
+      for (String prefix : GIT_LINE_PREFIXES) {
+        if (line.startsWith(prefix)) {
+          return LineType.GIT_LINE;
+        }
+      }
+    }
+    if (line.startsWith("@@") && line.lastIndexOf("@@") != 0) {
+      int pos = line.indexOf("@@", 2);
+      Matcher m = chunkHeaderRe.matcher(line.substring(0, pos + 2));
+      if (m.find()) {
+        return LineType.CHUNK_HEAD;
+      }
+    }
+    return LineType.UNKNOWN;
+  }
+
+  @VisibleForTesting
+  public static List<String> readFile(Path file) throws IOException {
+    return Lists.newArrayList(FileSystemUtils.readLines(file, StandardCharsets.UTF_8));
+  }
+
+  private static void writeFile(Path file, List<String> content) throws IOException {
+    FileSystemUtils.writeLinesAs(file, StandardCharsets.UTF_8, content);
+  }
+
+  private static void applyPatchToFile(List<String> patchContent, Path oldFile, Path newFile,
+      boolean isRenaming)
+      throws IOException, PatchFailedException {
+    // Neither of newFile and oldFile can be none if we should do a rename.
+    if (isRenaming && (oldFile == null || newFile == null)) {
+      String msg = "old file is " + (oldFile == null ? "not specified" : oldFile.getPathString());
+      msg += " and new file is " + (newFile == null ? "not specified" : newFile.getPathString());
+      throw new PatchFailedException("Cannot rename file without two valid file names, " + msg);
+    }
+
+    // The file we should read oldContent from.
+    Path inputFile = null;
+    if (oldFile != null && oldFile.exists()) {
+      inputFile = oldFile;
+    } else if (newFile != null && newFile.exists()) {
+      inputFile = newFile;
+    }
+
+    Patch<String> patch = DiffUtils.parseUnifiedDiff(patchContent);
+    // Does this patch look like adding a new file.
+    boolean isAddFile =
+        patch.getDeltas().size() == 1 && patch.getDeltas().get(0).getOriginal().getLines().isEmpty();
+
+    // If this patch is not adding a new file and we cannot find the input file, throw an error.
+    if (!isAddFile && inputFile == null) {
+      String msg = "old file is " + (oldFile == null ? "not specified" : oldFile.getPathString());
+      msg += " and new file is " + (newFile == null ? "not specified" : newFile.getPathString());
+      throw new PatchFailedException("Cannot find file to patch: " + msg);
```

Comment:
I guess the new file does not matter, also in most cases there would be no rename, only editing